topdown/walk.go

```diff
@@ -24,13 +41,22 @@ func evalWalk(_ BuiltinContext, operands []*ast.Term, iter func(*ast.Term) error
 }
 
 func walk(filter, path *ast.Array, input *ast.Term, iter func(*ast.Term) error) error {
-
 	if filter == nil || filter.Len() == 0 {
 		if path == nil {
 			path = ast.NewArray()
 		}
 
-		if err := iter(ast.ArrayTerm(ast.NewTerm(path.Copy()), input)); err != nil {
+		// TODO: why does this not work?
```

Comment:
TODO: let's find an answer before merging this 😉 

I think it doesn't work because defer doesn't go well with CPS. `return iter(...)` will mean that the defer runs when **everything** is evaluated successfully.

Reply:
> let's find an answer before merging this

Naturally :) I'll see if I can get something to work without defer ... but note that this actually _does_ work in the walkNoPath version... the few added unit tests on this confirms so, but even more so that `regal lint`works, as we're quite dependent on that.

Reply:
> I think it doesn't work because defer doesn't go well with CPS. return iter(...) will mean that the defer runs when everything is evaluated successfully.

This seems to align with what some println's tell me. However, I'm not sure why this would be an issue. Even if all path/value pairs aren't returned to the pool until after the whole walk is done, that would still be a substantial improvement over having to create all new items the next walk is invoked. BUT the YAML tests show some really odd behavior in the printed path/values vs the expected ones! But now that I try to run these changes using `regal lint bundle`, everything actually works.. and we use walk (even with path) _quite_ extensively... so I'm really really clueless as to what's going on right now 😅

Reply:
@srenatus I have now pushed my latest changes, that return items to the pool "manually" instead of via defer. I also added a pool for terms as they too can be reduced, and all in all this is another 4 million allocations down if successful. The OPA YAML tests related to walk are failing, where Regal just keeps going. I don't even know where to start 😄 But perhaps you'll come up with some ideas. I figured I'd push this so that we're on the same page.